pandas/tests/frame/methods/test_replace.py

```diff
@@ -1356,3 +1356,10 @@ def test_replace_replacer_dtype(self, replacer):
         result = df.replace({"a": replacer, "b": replacer})
         expected = pd.DataFrame([replacer])
         tm.assert_frame_equal(result, expected)
+
+    def test_replace_with_nan(self):
+        # GH30512
+        df = pd.DataFrame({"A": [np.nan, 3], "B": [1.096, 1.086], "C": [1, 2]})
```

Comment:
Does this only happen if `A` and `B` are both float dtypes before replacement?

Reply:
I think so, because then they're both part of the same FloatBlock and the whole thing gets cast to object, but I'll check

Reply:
It happens with other dtypes, e.g.
```
>>> data = pd.DataFrame({'A':np.nan,'B':2, 'C':1}, index=[0])
>>> data.dtypes
A    float64
B      int64
C      int64
dtype: object
>>> data.replace(to_replace={1:None}).dtypes
A    float64
B     object
C     object
```

Will update the test to cover all dtypes tomorrow

Reply:
@WillAyd have updated tests with other dtypes and reworded whatsnew